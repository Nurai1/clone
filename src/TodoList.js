import React, { useEffect } from 'react';
import { connect } from 'react-redux';

import * as actions from './actions';

export const TodoList = ({
  reducer,
  todos,
  sortingDetails,
  dispatch
}) => {
  let storeFromLocalStore = JSON.parse(localStorage.getItem("reducer"));

  useEffect(() => {
      localStorage.setItem("reducer", JSON.stringify(reducer));
  });

  const initiliazeSortFunc = function(sortItem) {
    if (sortingDetails.fromTop){
      return function(a,b){
          if (a[sortItem]>b[sortItem])
            return 1;
          else
            return -1;
        };
    }
    else {
      return function(b,a){
          if (a[sortItem]>b[sortItem])
            return 1;
          else
            return -1;
        };
    }
  }
  return (
    <div>
    <h3>Сортировка:</h3>
    <label htmlFor="">
      По тексту:
      <input type="radio" name="sortFunc"
        checked={sortingDetails.item==="text"}
        onChange={()=>{}}
        onClick={()=>{
          dispatch(actions.changeSortItem("text"));
        }}
      />
    </label>
    <label htmlFor="">
      По дате:
      <input type="radio" name="sortFunc"
        checked={sortingDetails.item==="date"}
        onChange={()=>{}}
        onClick={()=>{
          dispatch(actions.changeSortItem("date"));
        }}
      />
    </label>
    <label htmlFor="">
      Изменить направление:
      <input type="checkbox"
        checked={sortingDetails.fromTop}
        onChange={()=>{
          dispatch(actions.toggleSortOrder());
        }}
      />
    </label>
    <ul>
      {
        todos.sort(initiliazeSortFunc(sortingDetails.item)).map((todo)=>{
          let toDoId= todo.id;
          return (todo.filterText && todo.filterDate)?
            <li key={todo.id} >
              <input type="checkbox" onClick={()=>{dispatch(actions.toggleCompleteState(toDoId))}} />
              Текст: {todo.text.toString()} Дата: {todo.date.toString()}
              <input onClick={()=>{dispatch(actions.deleteToDo(toDoId))}} type="button" value="Удалить" />
            </li>:""
        })
      }
    </ul>
    </div>
  );
}

const mapStateToProps = (state) => ({
  reducer: state,
  todos: state.todos,
  sortingDetails: state.sortingDetails
});

const TodoListContainer = connect(
  mapStateToProps,
)(TodoList);

export default TodoListContainer;
