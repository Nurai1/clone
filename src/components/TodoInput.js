import React from 'react';
import { connect } from 'react-redux';

import {
  addCurrentDate,
  addCurrentText,
  addToDo,
  setInputErrorState
} from './../actions';

export const TodoInput = ({
  todoCurrentValues,
  inputErrorState,
  addCurrentText,
  addCurrentDate,
  addNewTodo,
  setInputErrorState
}) => {
  const addTodo = (e) => {
    e.preventDefault();
    if (!todoCurrentValues.date || !todoCurrentValues.text){
      setInputErrorState(true);
      return;
    }

    addNewTodo(todoCurrentValues.text, todoCurrentValues.date);
    setInputErrorState(false);
  }

  return (
    <form className="addForm" onSubmit={addTodo} >
      <h3>Добавьте задание</h3>
      <label className="addForm__text" htmlFor="">
        Введите текст задания:
        <input className={(!todoCurrentValues.text && inputErrorState)?"error":""}
          onChange={(e)=>addCurrentText(e.target.value)} type="text"
        />
      </label>
      <label className="addForm__date" htmlFor="">
        Введите дату:
        <input className={(!todoCurrentValues.date && inputErrorState)?"error":""}
          onChange={(e)=>addCurrentDate(e.target.value)} type="date"
        />
      </label>
      <input type="submit" value="Добавить" />
    </form>
  );
}

const mapStateToProps = (state) => ({
  todoCurrentValues: state.todoCurrentValues,
  inputErrorState: state.inputErrorState
})

const mapDispatchToProps = (dispatch) => ({
  addCurrentText: (value) => dispatch(addCurrentText(value)),
  addCurrentDate: (value) => dispatch(addCurrentDate(value)),
  addNewTodo: (text, date) => dispatch(addToDo(text, date)),
  setInputErrorState: (value) => dispatch(setInputErrorState(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(TodoInput);
