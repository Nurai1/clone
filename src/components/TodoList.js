import React from 'react';
import { connect } from 'react-redux';

import {
  changeSortItem,
  toggleSortOrder,
  toggleCompleteState,
  deleteToDo
} from './../actions';

export const TodoList = ({
  todos,
  sortingDetails,
  visibilityFilters,
  changeSortItem,
  toggleSortOrder,
  toggleCompleteState,
  deleteToDo
}) => {

  let currentTodos = [...todos];
  const initiliazeSortFunc = function(sortItem) {
    if (sortingDetails.fromTop){
      return function(a,b){
          if (a[sortItem]>b[sortItem])
            return 1;
          else
            return -1;
        };
    }
    else {
      return function(b,a){
          if (a[sortItem]>b[sortItem])
            return 1;
          else
            return -1;
        };
    }
  }

  const sortingDetailsHandler = function(item) {
    if (item === sortingDetails.item){
      toggleSortOrder();
    }
    changeSortItem(item);
  }

  return (
    <div>
    <h3>Todo List</h3>
      <input type="button"
        className="btn__sort_txt"
        value="Сортировка по тексту"
        onClick={()=>{
          sortingDetailsHandler("text");
        }}
      />
      <input type="button"
        className="btn__sort_date"
        value="Сортировка по дате"
        onClick={()=>{
          sortingDetailsHandler("date");
        }}
      />
      <input type="button"
        className="btn__sort_reset"
        value="Сброс сортировки"
        onClick={()=>{
          sortingDetailsHandler("");
        }}
      />
    <ul>
      {
        currentTodos.sort(initiliazeSortFunc(sortingDetails.item)).map((todo)=>{
          let toDoId= todo.id;
          return (todo.text.includes(visibilityFilters.text) && todo.date.includes(visibilityFilters.date))?
            <li key={todo.id} >
              <input type="checkbox"
              onChange={()=>{toggleCompleteState(toDoId)}}
              checked={todo.completed?"checked":""}
            />
              <span className={todo.completed?"task task_completed":"task"}>
                Текст: {todo.text.toString()} Дата: {todo.date.toString()}
              </span>
              <input onClick={()=>{deleteToDo(toDoId)}} type="button" value="Удалить" />
            </li>
            :null
        })
      }
    </ul>
    </div>
  );
}

const mapStateToProps = (state) => ({
  todos: state.todos,
  sortingDetails: state.sortingDetails,
  visibilityFilters: state.visibilityFilters
});

const mapDispatchToProps = (dispatch) => ({
  changeSortItem: (item) => dispatch(changeSortItem(item)),
  toggleSortOrder: () => dispatch(toggleSortOrder()),
  toggleCompleteState: (id) => dispatch(toggleCompleteState(id)),
  deleteToDo: (id) => dispatch(deleteToDo(id))
});

const TodoListContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoList);

export default TodoListContainer;
