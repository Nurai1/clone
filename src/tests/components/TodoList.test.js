import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer';

import { TodoList } from '../../components/TodoList';

Enzyme.configure({ adapter: new Adapter() });

function setup(props) {
  const wrapper = shallow(
    <TodoList {...props} />
  );

  return wrapper;
}

const props = {
  todos: [
    {
      id: 1,
      text: "create store",
      date: "22.02.2004",
      completed: false
    },
    {
      id: 2,
      text: "finished iting",
      date: "22.02.2003",
      completed: false
    }
  ],
  sortingDetails: {item:"text", fromTop: false},
  visibilityFilters: {text: "", date: ""},
  toggleSortOrder: jest.fn(),
  changeSortItem: jest.fn()
};

describe("Todo List Component", () => {

  it("capturing snapshot of component", () => {

    const renderedComponent = renderer.create(<TodoList {...props} />).toJSON();

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders without todo list", () => {
    const wrapper = setup({
      todos: [],
      sortingDetails: {item:"", fromTop: false},
      visibilityFilters: {text: "", date: ""}
    });

    expect(wrapper.find("ul").children().exists()).toEqual(false);
  });

  it("renders with todo list", () => {

    const wrapper = setup(props);

    expect(wrapper.find("ul").children()).toHaveLength(props.todos.length);
  });

  it("checks sorting function", () => {

    const wrapper = setup(props);
    const sortingFunction = function(sortItem, fromTop) {
      if (fromTop){
        return function(a,b){
            if (a[sortItem]>b[sortItem])
              return 1;
            else
              return -1;
          };
      }
      else {
        return function(b,a){
            if (a[sortItem]>b[sortItem])
              return 1;
            else
              return -1;
          };
      }
    };

    let todosCopy = [...props.todos];
    todosCopy = todosCopy.sort(
      sortingFunction(
        props.sortingDetails.item,
        props.sortingDetails.fromTop
  ));

    let renderedTodos = wrapper.find("li");
    renderedTodos.forEach((value, index) => {
      expect(value.key()).toBe(todosCopy[index].id+"");
    });

  });

  it("should show only one item at Todo list using visibility filter", () => {

    const localProps = {
      ...props,
      visibilityFilters: {text: "create", date: ""}
    };


    const wrapper = setup(localProps);
    const expectedLength = 1;

    expect(wrapper.find("li")).toHaveLength(expectedLength);
  });

  it(
    `should call toggleSortOrder props function, if try to click on the sorting
    button which won't change sorting item `, () => {

      const wrapper = setup(props);

      const txtButton = wrapper.find(".btn__sort_txt");
      txtButton.props().onClick();

      expect(props.toggleSortOrder.mock.calls.length).toBe(1);
  });

  it("should call changeSortItem props function with date parameter", () => {

    const wrapper = setup(props);

    const dateButton = wrapper.find(".btn__sort_date");
    // dateButton.props().onClick();
    dateButton.simulate("click");
    expect(props.changeSortItem).toHaveBeenCalledWith("date");
  });

});
