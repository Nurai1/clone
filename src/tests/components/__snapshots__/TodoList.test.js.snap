// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo List Component capturing snapshot of component 1`] = `
<div>
  <h3>
    Todo List
  </h3>
  <input
    className="btn__sort_txt"
    onClick={[Function]}
    type="button"
    value="Сортировка по тексту"
  />
  <input
    className="btn__sort_date"
    onClick={[Function]}
    type="button"
    value="Сортировка по дате"
  />
  <input
    className="btn__sort_reset"
    onClick={[Function]}
    type="button"
    value="Сброс сортировки"
  />
  <ul>
    <li>
      <input
        checked=""
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="task"
      >
        Текст: 
        finished iting
         Дата: 
        22.02.2003
      </span>
      <input
        onClick={[Function]}
        type="button"
        value="Удалить"
      />
    </li>
    <li>
      <input
        checked=""
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="task"
      >
        Текст: 
        create store
         Дата: 
        22.02.2004
      </span>
      <input
        onClick={[Function]}
        type="button"
        value="Удалить"
      />
    </li>
  </ul>
</div>
`;
